accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -2852,9 +2852,20 @@ To match a pattern `p` against a value `v`:
 
     5.  Match the head elements. For `i` from `0` to `h - 1`, inclusive:
 
-        1.  Extract the element value `e` by calling `v[i]`.
+        1.  Let `s` be the `i`th element subpattern.
 
-        2.  Match the `i`th element subpattern against `e`.
+        2.  If `s` is an identifier pattern whose name is `_` then do nothing
+            for this element.
+
+            *Wildcards are useful in list patterns to control the index that
+            latter element subpatterns access and to affect the length that the
+            pattern checks. When a user does that, we don't want them to pay a
+            runtime penalty for accessing list elements that the pattern won't
+            use anyway.*
+
+        3.  Else extract the element value `e` by calling `v[i]`.
+
+        4.  Match `s` against `e`.
 
     6.  If there is a matching rest element:
 
```

Comment:
I think we should avoid calling `sublist` if the rest element has no pattern, like `[var x, ..., var y]`. No need to create the sublist here, the `...` is just for filling out positional positions.

And because someone will probably end up writing `[var x, ... _, var y]` to mean the same thing, we may want to not create the sublist in that case either.

So.

>    1. If the rest element has no pattern, then do nothing for this element.
>    2. Otherwise let *s* be the rest element pattern after the `...`.
>    3. If *s* is an identifier pattern whose name is `_` then do nothing
>        for this element.
>    4. Otherwise ...

Reply:
> I think we should avoid calling sublist if the rest element has no pattern, like [var x, ..., var y]. No need to create the sublist here, the ... is just for filling out positional positions.

Agreed. That's why the proposal says "If there is a matching rest element" and not just "If there is a rest element".

> And because someone will probably end up writing [var x, ... _, var y] to mean the same thing, we may want to not create the sublist in that case either.

If, for some reason, the user chooses to write that, the only possible reason to do so over just `...` is because you _want_ it to call `sublist()`, so I think it makes sense to leave it as currently specced.